update library dependencies

michelou opened the pull request:
Updating dependency on Scala 2.13.5 to 2.13.10 requires the small change in source file [Weeder.scala](https://github.com/flix/flix/compare/master...michelou:flix:flix-lib-deps?expand=1#diff-ecb9df81b3be826080f39bf570a3a23e2a7587eb28d967085ac72d759be4ebda); otherwise no source change is needed.
**PS.** The update of libraries `json4s-*.jar` and `scalatest/scalactic.jar` is postponed as it requires further changes in source code.

magnus-madsen commented:
Hello @michelou and thanks for the PR :)

Two things:

- Could I ask you to add your name to `AUTHORS.md` to indicate that you agree to release your contributions under Apache 2.0?
- We usually upgrade dependencies very very carefully due to potential downstream effects. Hence I was actually only planning/expecting for Scala to be upgraded. Each other library we will upgrade separately and by need. If you could change that too, it would be great!

Thanks!

magnus-madsen commented:
I re-triggered the build because it took 21min. Usually it should take 11min.

magnus-madsen commented:
Thanks!

michelou commented:
@magnus-madsen Good news !

Starting with [**Scala 2.13.11-bin-c1a11b1-SNAPSHOT**](https://scala-ci.typesafe.com/ui/native/scala-pr-validation-snapshots/org/scala-lang/scala-library/2.13.11-bin-c1a11b1-SNAPSHOT) the performance issue associated with the Scala option [`-Ypatmat-exhaust-depth`](https://github.com/michelou/flix/blob/master/build.gradle#L45) is finally solved (see https://github.com/scala/scala/pull/10361) !

I will submit another PR once the final release of Scala **2.13.11** is available (as an "update" of https://github.com/flix/flix/pull/4896 which you had to revert on November 14, 2022).

Here are the build times of the [Flix checkout](https://github.com/michelou/flix/tree/master) of 2023-04-14 on a Windows 10 Pro machine :

| Scala&nbsp;version | Gradle&nbsp;build&nbsp;command | Execution&nbsp;time |
|:--------|:-------------------------|:-----------------------|
| [2.13.5](https://mvnrepository.com/artifact/org.scala-lang/scala-library/2.13.5)  | `gradle -b build.gradle clean jar` | 2m 3s, 2m 4s, 2m 10s, 2m 2s  |
| [2.13.10](https://mvnrepository.com/artifact/org.scala-lang/scala-library/2.13.10) | `gradle -b build-scala-2.13.10.gradle clean jar` |  7m 46s, 8m 15s, 8m 26s, 8m 23s |
| [2.13-11-SNAPSHOT](https://scala-ci.typesafe.com/ui/native/scala-pr-validation-snapshots/org/scala-lang/scala-library) | `gradle -b build-scala-2.13.11.gradle clean jar`  | 2m 8s, 2m 9s, 2m 8s  |

Needless to say, the status of the Flix test suite is **OK** for all versions (aka. "Tests: succeeded 1577, failed 0, canceled 0") : 

| Scala&nbsp;version | Gradle&nbsp;test&nbsp;command | Execution&nbsp;time |
|:--------|:-------------------------|:-----------------------|
| [2.13.5](https://mvnrepository.com/artifact/org.scala-lang/scala-library/2.13.5)  | `gradle -b build.gradle test` | 7m 11s  |
| [2.13.10](https://mvnrepository.com/artifact/org.scala-lang/scala-library/2.13.10) | `gradle -b build-scala-2.13.10.gradle test` |  7m 11s |
| [2.13-11-SNAPSHOT](https://scala-ci.typesafe.com/ui/native/scala-pr-validation-snapshots/org/scala-lang/scala-library) | `gradle -b build-scala-2.13.11.gradle test`  | 7m 9s  |

Finally I mention here my two changes in the case you're interested to build Flix yourself with my settings : 
1. The Gradle project file for Scala **2.13.10** looks as follows (***non working case***) : 
    ```
   F:\flix>diff build.gradle build-scala-2.13.10.gradle
   15,16c15,17
   <     implementation 'org.scala-lang:scala-library:2.13.5!!'
   <     implementation 'org.scala-lang:scala-reflect:2.13.5!!'
   ---
   >     implementation 'org.scala-lang:scala-compiler:2.13.10'
   >     implementation 'org.scala-lang:scala-library:2.13.10'
   >     implementation 'org.scala-lang:scala-reflect:2.13.10'
   43a45
   >             "-release", "11",
   ```
   The Gradle project file for Scala **2.13.11-SNAPSHOT** looks as follows : 
   ```
   F:\flix>diff build.gradle build-scala-2.13.11.gradle
   11a12,14
   >     maven {
   >         url = 'https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots'
   >     }
   15,16c18,23
   <     implementation 'org.scala-lang:scala-library:2.13.5!!'
   <     implementation 'org.scala-lang:scala-reflect:2.13.5!!'
   ---
   >     // tested Scala 2.13.11 snapshots :
   >     // 2.13.11-bin-c1a11b1-SNAPSHOT, 04-04-2023
   >     // 2.13.11-bin-8d45170-SNAPSHOT, 14-04-2023
   >     implementation 'org.scala-lang:scala-compiler:2.13.11-bin-8d45170-SNAPSHOT'
   >     implementation 'org.scala-lang:scala-library:2.13.11-bin-8d45170-SNAPSHOT'
   >     implementation 'org.scala-lang:scala-reflect:2.13.11-bin-8d45170-SNAPSHOT'
   43a51
   >             "-release", "11",
   ```
3. We had to update **4** Scala source files which contain the now reserved word **`enum`** for variables/parameters (we renamed those occurrences to `enum0`) : 
    ```
    F:\flix>git status -uno
    On branch master
    Your branch is up to date with 'origin/master'.
    &nbsp;
    Changes not staged for commit:
      (use "git add <file>..." to update what will be committed)
      (use "git restore <file>..." to discard changes in working directory)
            modified:   main/src/ca/uwaterloo/flix/language/errors/DerivationError.scala
            modified:   main/src/ca/uwaterloo/flix/language/phase/Kinder.scala
            modified:   main/src/ca/uwaterloo/flix/language/phase/Resolver.scala
            modified:   main/src/ca/uwaterloo/flix/language/phase/inference/RestrictableChooseInference.scala
    ```

magnus-madsen commented:
Great news. Thanks for keeping us posted 😊
